org.eclipse.xtext/src/org/eclipse/xtext/xtext/XtextValidator.java

```diff
@@ -1234,5 +1235,45 @@ public void checkOppositeReferenceUsed(Assignment assignment) {
 			}
 		}
 	}
-
+	
+	@Check
+	public void checkAnnotations(AbstractRule rule) {
+		final List<String> validNames = AnnotationNames.VALID_ANNOTATIONS_NAMES;
+		for (Annotation a : rule.getAnnotations()) {
+			if (!validNames.contains(a.getName())) {
+				error("This annotation type is not supported.", a, XtextPackage.Literals.ANNOTATION__NAME);
+			}
+		}
+	}
+	
+	@Check
+	public void checkOverridingRule(AbstractRule rule) {
+		final String name = rule.getName();
+		
+		final boolean isOverride =
+				rule.getAnnotations().stream().anyMatch(e -> AnnotationNames.OVERRIDE.equals(e.getName()));
+		
+		for (Grammar g : GrammarUtil.getGrammar(rule).getUsedGrammars()) {
+			final AbstractRule r = GrammarUtil.findRuleForName(g, rule.getName());
+			
+			if (r != null) {
+				if (!isOverride) {
+					warning("This rule overrides " + name + " in " + GrammarUtil.getGrammar(r).getName() + ".", rule,
```

Comment:
* The warning message should also indicate what to do about it:
`This rule overrides Foo in Bar and thus should be annotated with @Override".

* please add an Issue Code, so that later a QuickFix can be provided.